config: Don't spam warnings about reference type configs

c00w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145800

Summary:
https://github.com/pytorch/pytorch/issues/145755

The is_dynamic check for reference types was subtly broken, causing log spam
after it was accessed

Added an explicit type for is_default for reference types to make sure this
behaviour is correct

c00w committed:
config: Don't spam warnings about reference type configs

Summary:
https://github.com/pytorch/pytorch/issues/145755

The is_dynamic check for reference types was subtly broken, causing log spam
after it was accessed

Added an explicit type for is_default for reference types to make sure this
behaviour is correct

[ghstack-poisoned]

c00w committed:
Update on "config: Don't spam warnings about reference type configs"


Summary:
https://github.com/pytorch/pytorch/issues/145755

The is_dynamic check for reference types was subtly broken, causing log spam
after it was accessed

Added an explicit type for is_default for reference types to make sure this
behaviour is correct

[ghstack-poisoned]

c00w commented:
Also confirmed repro from underlying issue is fixed (no more spam seen).

ezyang commented:
The changes in this PR are reasonable but I still don't understand why we are warning at all here? (Even warn once seems too much?)

c00w commented:
@ezyang  - It's what the old code did, Let me just remove the logging (I don't feel strongly about it).

c00w committed:
Update on "config: Don't spam warnings about reference type configs"


Summary:
https://github.com/pytorch/pytorch/issues/145755

The is_dynamic check for reference types was subtly broken, causing log spam
after it was accessed

Added an explicit type for is_default for reference types to make sure this
behaviour is correct

[ghstack-poisoned]

eellison commented:
@ezyang i added it originally because first we would fail when trying to create a repro when `fx_post_grad_pass`, now we are warning that we are ignoring it because we cant serialize. 

given that this is a developer-facing warning (or supposed to be) I think it makes sense to let developer know that we were unable to serialize the custom pass, or other config option. although here it was firing errantly

eellison commented on torch/utils/_config_module.py:
```diff
@@ -444,11 +443,12 @@ def _is_default(self, name: str) -> bool:
+
+        unset = config_val.user_override is _UNSET_SENTINEL
+        # Handle reference types specially to avoid spammy warnings
+        if isinstance(config_val.default, (list, set, dict)):
+            unset = unset or config_val.user_override == config_val.default
```
is there a reason why we wouldn't default to value equality more generally ?

c00w replied:
We can, for all types originally just comparing with _UNSET_SENTINEL was slightly better., but then we had to cludge around with reference semantics.

The one trick is we only allow a lot of overrides if the value is not set by the user (which breaks for reference types, but they also don't allow overrides anyway).

ezyang commented:
Yeah, I think my point is, if it's supposed to only trigger when a developer is doing something, then why is it triggering here? (I guess TORCH_TRACE maybe?)

c00w commented:
@ezyang  - The bug was that we always logged it if anyone accessed a reference field (which is what the is_default fix is about). Since to prevent unintended mutations to underlying defaults, if the type is a reference type, we would copy it before we returned it.

c00w commented:
@pytorchbot merge

c00w committed:
Update on "config: Don't spam warnings about reference type configs"


Summary:
https://github.com/pytorch/pytorch/issues/145755

The is_dynamic check for reference types was subtly broken, causing log spam
after it was accessed

Added an explicit type for is_default for reference types to make sure this
behaviour is correct

[ghstack-poisoned]

c00w committed:
Update on "config: Don't spam warnings about reference type configs"


Summary:
https://github.com/pytorch/pytorch/issues/145755

The is_dynamic check for reference types was subtly broken, causing log spam
after it was accessed

Added an explicit type for is_default for reference types to make sure this
behaviour is correct

[ghstack-poisoned]

c00w commented:
@pytorchbot merge

c00w commented:
@pytorchbot merge -f "windows failure is infra + unrelated"
